File: python/clarian/webhooks.py
"""Webhook signature helpers (HMAC-SHA256 hex over timestamp.body)."""

from __future__ import annotations

import hashlib
import hmac
import time
from datetime import datetime, timezone
from email.utils import parsedate_to_datetime
from typing import Mapping, Sequence, Union

HEADER_SIGNATURE = "X-Clarian-Signature"
HEADER_TIMESTAMP = "X-Clarian-Timestamp"
HEADER_EVENT = "X-Clarian-Event"
HEADER_DELIVERY_ID = "X-Clarian-Delivery-Id"
HEADER_IDEMPOTENCY_KEY = "X-Clarian-Idempotency-Key"
HEADER_ATTEMPT = "X-Clarian-Attempt"

DEFAULT_TOLERANCE_SECONDS = 300

Headers = Mapping[str, Union[str, Sequence[str], None]]


def _as_bytes(payload: bytes | str) -> bytes:
    if isinstance(payload, bytes):
        return payload
    return payload.encode("utf-8")


def _as_str(payload: bytes | str) -> str:
    if isinstance(payload, str):
        return payload
    return payload.decode("utf-8")


def sign_payload(secret: str, timestamp: str, payload: bytes | str) -> str:
    """Return hex HMAC-SHA256 of ``f\"{timestamp}.{body}\"`` using secret."""
    body = _as_str(payload)
    mac = hmac.new(
        secret.encode("utf-8"),
        f"{timestamp}.{body}".encode("utf-8"),
        hashlib.sha256,
    )
    return mac.hexdigest()


def _parse_timestamp(timestamp: str) -> float:
    try:
        # ISO 8601 / RFC3339 (with or without Z)
        ts = timestamp.replace("Z", "+00:00")
        dt = datetime.fromisoformat(ts)
        if dt.tzinfo is None:
            dt = dt.replace(tzinfo=timezone.utc)
        return dt.timestamp()
    except ValueError:
        pass
    try:
        return float(timestamp)
    except ValueError:
        pass
    try:
        return parsedate_to_datetime(timestamp).timestamp()
    except (TypeError, ValueError, IndexError):
        raise ValueError(f"invalid webhook timestamp: {timestamp!r}") from None


def verify_signature(
    payload: bytes | str,
    timestamp: str,
    signature: str,
    secret: str,
    tolerance_seconds: int = DEFAULT_TOLERANCE_SECONDS,
) -> bool:
    """Verify HMAC-SHA256 signature and timestamp freshness.

    Signature is hex(HMAC-SHA256(secret, timestamp + "." + body)).
    Returns False on missing inputs, bad signature, or stale/future timestamp.
    """
    if not secret or not signature or not timestamp:
        return False

    try:
        event_ts = _parse_timestamp(timestamp)
    except ValueError:
        return False

    age = abs(time.time() - event_ts)
    if age > tolerance_seconds:
        return False

    expected = sign_payload(secret, timestamp, payload)
    return hmac.compare_digest(expected, signature)


def extract_headers(headers: Headers) -> tuple[str, str]:
    """Return ``(timestamp, signature)`` from Clarian delivery headers.

    Reads ``X-Clarian-Timestamp`` and ``X-Clarian-Signature`` (case-insensitive).
    Missing headers yield empty strings.
    """

    def get(name: str) -> str:
        lower = name.lower()
        for key, value in headers.items():
            if key.lower() != lower:
                continue
            if value is None:
                return ""
            if isinstance(value, (list, tuple)):
                return str(value[0]) if value else ""
            return str(value)
        return ""

    return get(HEADER_TIMESTAMP), get(HEADER_SIGNATURE)

File: ruby/lib/clarian/webhooks.rb
# frozen_string_literal: true

require "openssl"
require "time"

module Clarian
  # Webhook signature helpers (HMAC-SHA256 hex over timestamp.body).
  module Webhooks
    HEADER_SIGNATURE = "X-Clarian-Signature"
    HEADER_TIMESTAMP = "X-Clarian-Timestamp"
    HEADER_EVENT = "X-Clarian-Event"
    HEADER_DELIVERY_ID = "X-Clarian-Delivery-Id"
    HEADER_IDEMPOTENCY_KEY = "X-Clarian-Idempotency-Key"
    HEADER_ATTEMPT = "X-Clarian-Attempt"

    DEFAULT_TOLERANCE_SECONDS = 300

    module_function

    # Return hex HMAC-SHA256 of "#{timestamp}.#{body}" using secret.
    def sign_payload(secret:, timestamp:, payload:)
      body = payload.is_a?(String) ? payload : payload.to_s
      OpenSSL::HMAC.hexdigest("SHA256", secret.to_s, "#{timestamp}.#{body}")
    end

    # Verify HMAC-SHA256 signature and timestamp freshness.
    # Returns false on missing inputs, bad signature, or stale/future timestamp.
    def verify_signature(payload:, timestamp:, signature:, secret:, tolerance_seconds: DEFAULT_TOLERANCE_SECONDS)
      return false if secret.to_s.empty? || signature.to_s.empty? || timestamp.to_s.empty?

      begin
        event_ts = parse_timestamp(timestamp)
      rescue ArgumentError
        return false
      end

      age = (Time.now.to_f - event_ts).abs
      return false if age > tolerance_seconds

      body = payload.is_a?(String) ? payload : payload.to_s
      expected = sign_payload(secret: secret, timestamp: timestamp, payload: body)
      return false if expected.bytesize != signature.bytesize

      OpenSSL.secure_compare(expected, signature)
    end

    # Read Clarian delivery headers (case-insensitive). Missing values become empty strings.
    # Returns a Hash with keys: signature, timestamp, event, delivery_id, idempotency_key, attempt.
    def extract_headers(headers)
      return empty_headers if headers.nil?

      lower = {}
      headers.each do |key, value|
        next if key.nil?

        lower[key.to_s.downcase] = first_value(value)
      end

      {
        signature: lower[HEADER_SIGNATURE.downcase] || "",
        timestamp: lower[HEADER_TIMESTAMP.downcase] || "",
        event: lower[HEADER_EVENT.downcase] || "",
        delivery_id: lower[HEADER_DELIVERY_ID.downcase] || "",
        idempotency_key: lower[HEADER_IDEMPOTENCY_KEY.downcase] || "",
        attempt: lower[HEADER_ATTEMPT.downcase] || ""
      }
    end

    def empty_headers
      {
        signature: "",
        timestamp: "",
        event: "",
        delivery_id: "",
        idempotency_key: "",
        attempt: ""
      }
    end
    private_class_method :empty_headers

    def first_value(value)
      return "" if value.nil?
      return value.first.to_s if value.is_a?(Array)
      value.to_s
    end
    private_class_method :first_value

    def parse_timestamp(timestamp)
      # Unix epoch (seconds as string or number-like)
      if timestamp.match?(/\A-?\d+(\.\d+)?\z/)
        return timestamp.to_f
      end

      Time.iso8601(timestamp).to_f
    rescue ArgumentError
      raise ArgumentError, "invalid webhook timestamp: #{timestamp.inspect}"
    end
    private_class_method :parse_timestamp
  end
end
